### ISSUE TITLE
Feature request: validateScopes = true

### ISSUE THREAD

**Issue by duongphuhiep** [DevTeam/Pure.DI#69]
It would be nice if Pure.DI can warn us about "unusual" composition. For eg:

```C#
class Service(IRequest request) : IService;

DI.Setup()
            .Bind().As(Lifetime.Scoped).To<Request>()
            .Bind().As(Lifetime.Singleton).To<Service>()
            .Root<IService>(nameof(Service));
```

It is dangerous that a [Singleton] references to a  [Scope], [PerResolve], [PerBlock].. because they might potentially holding State (of the Scope, the Block or something else).

When you inject a scoped service into a singleton, the scoped service effectively becomes a singleton as well. This means that the scoped service instance will persist beyond the intended scope (like a web request), leading to potential issues such as:

* State Persistence: Scoped services might hold state that is intended to be request-specific. If they are injected into singletons, that state could persist across multiple requests, leading to data leakage or inconsistent behavior.

* Resource Management: Scoped services are typically disposed of at the end of their scope. If a singleton holds a reference to a scoped service, it won't be disposed of correctly, potentially causing memory leaks

it is why Microsoft DI framework have [validateScopes = true](https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection#scope-validation) to ensure that: a Singleton should only reference to (or should  only depend on)
- other Singletons
- or a Transient (which will be promote to Singleton). A transient usually doesn't contain any state, so it is fine.

[Transient]: https://github.com/DevTeam/Pure.DI/blob/master/readme/transient.md
[Singleton]: https://github.com/DevTeam/Pure.DI/blob/master/readme/singleton.md
[PerResolve]: https://github.com/DevTeam/Pure.DI/blob/master/readme/perresolve.md
[PerBlock]: https://github.com/DevTeam/Pure.DI/blob/master/readme/perblock.md
[Scope]: https://github.com/DevTeam/Pure.DI/blob/master/readme/scope.md

**Comment by NikolayPianikov** [comment]
@duongphuhiep 
1. Why can't Singleton create a session for itself and not use a Scoped dependency in it? For example such a “default” session.  On disposing, Singleton can dispose of the “default” session.
2. Singleton can create a scope, inject a Scoped dependency, use this dependency, but not save it to itself and free the scope.
3. As for PerResolve, PerBlock. Sure Singleton can use and even save them for future use. Since PerResolve, PerBlock are very similar to Tranisent in nature.

The lifetime of any dependency is defined as the longest possible lifetime of it and the object where it is injected, provided of course that it is preserved in its fields. I mean:

Traget lifetime + Dependency lifiteme = Actual dependency lifetime
Singleton + Scoped = Singleton
Scoped + Singleton = Singleton
Singleton + Transient = Singleton
Scoped + Transient = Scoped
Scoped + PerBlock = Scoped
...

I'm afraid there will be a lot of warnings and not always on the point. It might be worth adding a warning for the Singleton<-Scoped case, but with an option to disable it.

**Comment by duongphuhiep** [comment]
It is not only worth adding a warning for the **Singleton<-Scoped case**, I think it is even worth to prohibit it completely!

The typical use case I've seen is that a SingletonService depends on a HttpContext (scoped lifetime).
the HttpContext holds the Username ("Peter") => so the SingletonService authorized "Peter" to do stuff

When the second Http Request arrives with other Username ("Anonymous"), the SingletonService still hold the first HttpContext and so continue do things under the name of "Peter".. => user "Anonymous" sucessfully does stuff under the name of "Peter"

if PureDI can prevent this situation then it would be nice.

Note: My example might no longer accurate today with the HttpContextAccessor which used AsyncLocal, I only wanted to show the point.

**Comment by NikolayPianikov** [comment]
@duongphuhiep yeah, that's probably best. Thanks for the ticket. Will be done in the next release.

**Comment by NikolayPianikov** [comment]
Fixed in 2.1.32
